BoxE scoring function

AlCatt91 opened the pull request:
Added BoxE scoring function. This is the same implementation used in PK (a bit different from the one in the paper, but it looks to perform better). 
Also, some refactoring in the definition of `embedding_size` in the scoring functions, to make it more rigorous/better align it with intuition.

AlCatt91 commented:
Sorry for adding to this PR, I did some refactoring on how embedding tables are initialized, to be able to support standard `torch.nn.init` initializers and add more granularity for when (like in BoxE) you want to allocate multiple tensors (each of which needs to be initialized separately) per entity/relation.

danjust commented on besskge/scoring.py:
```diff
@@ -602,6 +638,254 @@ def score_tails(
+        )
+        self.apply_tanh = apply_tanh
+        self.dist_func_per_dim = dist_func_per_dim
+        self.soft = 1e-6
```
Should we have this as a param that defaults to 1e-6? And maybe call it `eps` or similar - would be more intuitive to me :)

danjust commented on besskge/scoring.py:
```diff
@@ -602,6 +638,254 @@ def score_tails(
+        )
+        # Rescale by box_size
+        width_ht = width_ht * (
+            torch.tensor(1.0, dtype=torch.float32, device=width_ht.device)
```
is it necessary to explicitly create these tensors (since it's just single floats)? Won't `torch.nn.functional.elu(box_size.unsqueeze(-1) + 1` perform identically?

danjust replied:
Ah, of course that makes sense - I'm not aware of a better way, I just ignore those warnings... Thanks 👍

danjust commented on besskge/scoring.py:
```diff
@@ -602,6 +638,254 @@ def score_tails(
+        if self.apply_tanh:
+            width_ht = torch.tanh(width_ht)
+        width_plus1_ht = (
+            torch.tensor(1.0, dtype=torch.float32, device=width_ht.device) + width_ht
```
see above

danjust commented on besskge/scoring.py:
```diff
@@ -602,6 +638,254 @@ def score_tails(
+            torch.tensor(1.0, dtype=torch.float32, device=width_ht.device) + width_ht
+        )
+        k_ht = (
+            torch.tensor(0.5, dtype=torch.float32, device=width_ht.device)
```
see above

danjust commented on tests/test_bess.py:
```diff
@@ -22,7 +22,7 @@
     TripleBasedShardedNegativeSampler,
 )
-from besskge.scoring import ComplEx, TransE
+from besskge.scoring import DistMult, TransE
```
Would it be worth unit testing a single fwd execution of the scoring functions against their pykeen counterpart? (Not necessarily in this PR!)

AlCatt91 replied:
Good point, I did want to have some unit tests for scoring functions but wasn't sure how to structure them. Comparing against pykeen might be the most logical way, I agree. I'll have a separate PR for that.

danjust reviewed (approved):
Looks really great, just very minor comments and questions

AlCatt91 commented:
Thanks Daniel for the review! I implemented the changes you suggested and fixed a small mistake for the way boxe was implemented with `apply_tanh=True` - please have a final check when you have time :)

danjust commented on besskge/scoring.py:
```diff
@@ -763,20 +769,22 @@ def boxe_score(
         )
         if self.apply_tanh:
-            width_ht = torch.tanh(width_ht)
+            width_ht = torch.tanh(width_ht) * torch.tensor(
```
I'm probably missing something, but is this not different from PyKEEN where tanh is applied to the box edges?

AlCatt91 replied:
Correct, mathematically it's not exactly the same implementation. I would like to avoid having to materialize the box edges if possible: when not using `tanh` it's simple to do (while pykeen has wasted computations there), here it's a bit of a workaround. In the tests I ran it still seems to perform similarly to pykeen, although if we want to have the exact same results we would need to do it differently.

danjust replied:
Would it be closer to PyKeen if `in_or_out` would always be computed pre-tanh? Due to monotonicity of tanh that should be equivalent to the case where tanh is applied to the box edges, whereas applying tanh to the box width can change which points fall into the box

AlCatt91 replied:
> So the factor 2 comes in with the approximation `tanh(x)~x` for small x?

The factor of 2 was just so the final range of `widht_ht` was [0,2], i.e. the same of what we would have if we took the difference of the tanh of the upper and lower edge.

> Would it be closer to PyKeen if `in_or_out` would always be computed pre-tanh?  Due to monotonicity of tanh that should be equivalent to the case where tanh is applied to the box edges, whereas applying tanh to the box width can change which points fall into the box

That is an excellent idea and I gave it a try with high expectations, but turns out to be actually slightly detrimental on final MRR (not sure why).

In the end I resorted to compute the upper and lower edges and do tanh like pykeen: impact on throughput is negligible and it does perform a bit better on MRR. I also moved some computations inside the `boxe_score` auxiliary function to minimize similar code across `score_triple, score_head, score_tail`. Hopefully this is the last iteration, but do tell if you see anything that could be still done better!

danjust reviewed (approved):
That's fantastic, thanks Alberto. And sorry for the tanh nitpicking...

AlCatt91 commented:
No worries at all, your remarks were 100% on point :) Thank you for the review!
